daml/ContingentClaims/Lifecycle.daml

```diff
@@ -60,106 +74,136 @@ lifecycle spot
       . fmap (fmap sequence)
       . fmap sequence
       . fmap
-      . fmap (fmap (fmap embed))
-      . fmap (fmap (tsidE : F t a (Either (C t a) (C t a)) -> Either (C t a) (F t a (C t a)) ))
       . fmap lift
-      $ acquire' spot
+      $ acquire' spot today
     distE = distApo
-     -- This is degenerate if you have e.g. `And Left Right`, but that should never happen in practice.
-    tsidE f = case sequence_ f of
-      Left _ -> Left (embed (fmap foldE f))
-      Right _ -> Right (fmap foldE f)
-    foldE (Left x) = x
-    foldE (Right x) = x
 
--- | Evaluate observables and skip branches for which predicates don't hold.
--- Consumes `When` and `Cond` when appropriate, leaving their subtrees.
+-- | Evaluate observables and skip branches for which predicates don't hold, calculate the acquisition time of the claim's sub-trees.
+-- Consume `Cond` nodes, leaving the relevant sub-tree.
+-- Replace `When cond c` nodes with `When (TimeGte t) c` if the condition is non-deterministic and evaluates to True.
 -- This is useless on its own. Composed with other functions, it adds laziness.
-acquire' : (Ord t, Eq a, Monad m)
+acquire' : (Ord t, Eq a, CanAbort m)
   => (a -> t -> m Decimal)
-  -> C t a -> Kleisli m t (F t a (Either (C t a) (C t a)))
-acquire' spot (When obs c) = do
-  predicate <- compare spot obs
-  if predicate then acquire' spot c else pure $ WhenF obs (Left c)
-acquire' spot (Cond obs c c') = do
-  predicate <- compare spot obs
-  if predicate then acquire' spot c else acquire' spot c'
-acquire' spot (Anytime obs c) = do
-  predicate <- compare spot obs
-  pure $ AnytimeF obs if predicate then Right c else Left c
-acquire' _ other = pure $ Right <$> project other
+  -- ^ function to evaluate observables 
+  -> t
+  -- ^ the today date
+  -> (AcquisitionTime t, C t a)
+  -- ^ input claim and its acquisition time
+  -> m (Either (C t a) (AcquisitionTime t, C t a))
+acquire' spot t (at, (When obs c)) = do
+  predicate <- compare spot obs t
+  if predicate then
+    case obs of
+      TimeGte s -> pure $ Right . (max at (Time s), ) $ When obs c
+      other -> pure $ Right . (Time t, ) $ When (TimeGte t) c
+      -- ^ if the condition `cond` is non-deterministic, we assume that we are lifecycling at a time `t` corresponding to the first instant the condition becomes True
+    else
+      pure . Left $ When obs c
+acquire' spot today (at,(Cond obs c c')) = do
+  predicate <- case at of
+    Time s -> compare spot obs s
+    MinusInfinity -> abort "Found Cond node with undefined acquisition time"
+  if predicate 
+    then acquire' spot today (at,c) 
+    else acquire' spot today (at,c')
+acquire' spot today (_,c@(Anytime obs _)) = do
```

Comment:
So this is insightful: I was struggling to see why we needed to have the acquistion time be separate from `today`.
Are we really sure we need `today` anymore though? I thought you said that the acquisition date was taken as the `max` of the outer contract, and the predicate? Would that make it possible to get rid of `today` altogether?

Reply:
In the case of a non-deterministic predicate (e.g. `When ( S(t) <= 50 )`) we can't easily determine the acquisition time of the inner contract, so the current best guess is to verify if the predicate holds on the `today` date